packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts

```diff
@@ -1,20 +1,82 @@
-const debug = require("debug")("compile:compilerSupplier");
-const requireFromString = require("require-from-string");
-const fs = require("fs");
-const originalRequire = require("original-require");
-const axios = require("axios").default;
-const semver = require("semver");
-const solcWrap = require("solc/wrapper");
-const LoadingStrategy = require("./LoadingStrategy");
-
-class VersionRange extends LoadingStrategy {
+import debugModule from "debug";
+const debug = debugModule("compile:compilerSupplier");
+
+import requireFromString from "require-from-string";
+import originalRequire from "original-require";
+import axios from "axios";
+import semver from "semver";
+import solcWrap from "solc/wrapper";
+import { Cache } from "../Cache";
+import { observeListeners } from "../observeListeners";
+import { NoVersionError, NoRequestError } from "../errors";
+
+export class VersionRange {
+  private config: {
+    events: any; // represents a @truffle/events instance, which lacks types
+    compilerRoots: string[];
+  };
+
+  private cache: Cache;
+
+  constructor(options) {
+    const defaultConfig = {
+      compilerRoots: [
+        // NOTE this relay address exists so that we have a backup option in
+        // case more official distribution mechanisms fail.
+        //
+        // currently this URL just redirects (302 Found); we may alter this to
+        // host for real in the future.
+        "https://relay.trufflesuite.com/solc/bin/",
+        "https://solc-bin.ethereum.org/bin/",
+        "https://ethereum.github.io/solc-bin/bin/"
+      ]
+    };
+    this.config = Object.assign({}, defaultConfig, options);
+
+    this.cache = new Cache();
+  }
+
+  async load(versionRange: string) {
+    const rangeIsSingleVersion = semver.valid(versionRange);
+    if (rangeIsSingleVersion && this.versionIsCached(versionRange)) {
+      return this.getCachedSolcByVersionRange(versionRange);
+    }
+
+    try {
+      return await this.getSolcFromCacheOrUrl(versionRange);
+    } catch (error) {
+      if (error.message.includes("Failed to complete request")) {
+        return this.getSatisfyingVersionFromCache(versionRange);
+      }
+      throw error;
+    }
+  }
+
+  async list() {
+    const data = await this.getSolcVersions();
+    const { latestRelease } = data;
+
+    const prereleases = data.builds
+      .filter(build => build["prerelease"])
+      .map(build => build["longVersion"]);
+
+    // ensure releases are listed in descending order
+    const releases = semver.rsort(Object.keys(data.releases));
```

Comment:
@haltman-at notes [here](https://github.com/trufflesuite/truffle/pull/4222#pullrequestreview-728040585) that we may be able to make use of the semver package's `maxSatisfying` function